### DISCUSSION TITLE
Do we need mappers?

### DISCUSSION THREAD

**Discussion by nmanero** [nmanero/ddd-crud#9] [General]
Hi,
In the #6 PR Elias sent many things and one of them was mappers. Do we really need them?
I don't like them very much, I would prefer to keep the toDTO... methods in the entities or in the domain classes more than having a mapper. What do you think?

**Comment by sergiogragera** [comment]
Some time ago I had the same opinion as @nmanero, but after reading [this](https://www.carlosble.com/2015/05/dto-vs-domain-models/?lang=es) I understood that it’s not a good idea coupling the entities to dto.

> `var dto = domainModel.ToDto(); // model to dto`
`var model = DomainModel.From(dto); // model from dto`

> But as many people commented on this post and as we saw later, it’s not a good idea because we are coupling the domain models to the way data is exported and sent out of the hexagon. Coupling is unavoidable...

**Reply by nmanero** [reply]
According to that, the adapter to convert from dto to domain it's ok, as you only call the new Domain(...) but toDto method, you need to fill the domain model with public getters, so I think I still prefer the domain.toDTO or maybe something more generic like domain.toHashMap

**Reply by sergiogragera** [reply]
I think you should keep the toDTO method in the adapter, because you may want to have several DTOs, one to expose information to the front and another to expose information to the persistence system, for example.

So in the domain model you can have toHashMap method and in the adapter have toDTO method and toPersistence method. But I think that toHashMap method needs to expose as many properties as you need in persistence DTO and for this reason I have getters in domain model and I have an adapter to expose only the information that you need in each DTO.

**Reply by elias480** [reply]
I think both conversion methods may be will be required at some point in a possible application.

The idea behind the mapper is try to apply solid principles and make smaller clases with less responsibilities.

I think I have not understood something. The purpose of toHashMap would not be the same of toDTO?
As @sergiogragera says many DTOs can be required, I like the idea of group all of them in the adapter/mapper class.

**Reply by nmanero** [reply]
The thing is that I don't like to add a lot of getters in the domain, that's why I was proposing to have the toDTO methods within the class.
As you both say, there can be may different toDTOs depending on the number of classes that we have and that is why I suggested to have a toHashMap method inside the domain, to return a key,value map with the properties,values of the domain. Then, we don't need to have a bunch of getters and we only return them in a hashmap. I don't like it very much but I'm not sure which option is worse.
I guess the domain constructor is enough, so we don't need the 
`DomainModel.From(dto)`
but in regard of 
`domainModel.ToDto();`
I'm not sure

**Reply by sergiogragera** [reply]
But what's the problem with getters? 

I think the idea is to use encapsulation only to mutate the state of the properties following a logic (we don't need to have a bunch of **setters**). 

By the other hand as @elias480 says, hiding information is responsability of mapper (I prefer to call it mapper than adapter). Then if you don't want to expose a property to the front ~~the service layer will return a DTO without this property and everyone who uses the service~~ the controller layer will return a DTO without this property and everyone who uses this endpoint will not know this property.

**Reply by nmanero** [reply]
It is not a problem with getters but if you have a public class with public getters anyone can access them and I wanted to see if there is any option. I would like to have only the constructors public, but maybe there is no other better option

**Reply by sergiogragera** [reply]
The persistence layer has to know this properties, the security layer will also need to know this properties, the layer of... reading properties is not a problem, the problem is being able to access and manipulate properties without a business rule: get mutable list (this get is the evil). But you can build a get of immutable list from your mutable property.

I think data hiding (I understand this is very important) is the responsibility of the controller, not a domain model representing a state.

**Reply by nmanero** [reply]
I know we need it. I just wanted to discuss about having getters or other way of accessing that information

**Reply by elias480** [reply]
Then after the discussion, what approach should we follow now?

**Reply by sergiogragera** [reply]
I think we should use data mappers like [this](https://khalilstemmler.com/articles/typescript-domain-driven-design/repository-dto-mapper/#Data-Mappers). We could have `abstract class` to extend **custom** mappers.

```
export abstract class Mapper<DomainEntityOrValueObject> {
   public static toDomain (raw: any): T;
   public static toDTO (t: T): DTO;
   public static toPersistence (t: T): any;
}
```

**Reply by nmanero** [reply]
I added a specific issue for this #11

**Reply by elias480** [reply]
Perfect, I'll resolve #5 also renaming valueObject by domain, it will be ok?

**Reply by nmanero** [reply]
feel free
